### ISSUE TITLE
Proxy Enhancement:  Cache size setting

### ISSUE THREAD

**Issue by sbulen** [SimpleMachines/SMF#4449]
After watching my (2.0 proxy) cache grow over 600MB of files a day, I realized we probably need some cache management...

I propose a user-configurable cache size setting, that is enforced via a Daily maintenance task.   (It grows too quickly for Weekly maintenance, on my site anyway...)

This applies to 2.0 & 2.1.

**Comment by sbulen** [comment]
This may be helpful, it's the one I'm now using:
https://github.com/sbulen/sjrbTools/blob/master/proxy-maint-cron.php

**Comment by albertlast** [comment]
i see this issue as "feature request" and
i notice no one develop this features.
So it's relevant for rc1?

**Comment by jdarwood007** [comment]
This would be a borderline feature request @albertlast.  It is a feature request in the sense we are adding new functionality.  However it also is a bug in the sense that the software is not managing its caching properly resulting in bloated storage usage.  @colinschoen can approve or deny the request to make this a addition or not sense there is doubt.  But he also added it as a milestone for RC1, so I assume he is on board with fixing this before releasing RC1.

**Comment by colinschoen** [comment]
This should be included in RC1. Without this feature and removing the SSL only for auth feature large boards are unable to realistically use SSL.

**Comment by albertlast** [comment]
ssl is not a real issue anymore today.
And smf 2.0.x got also no cache management?

**Comment by colinschoen** [comment]
What do you mean SSL is not a real issue anymore?

**Comment by albertlast** [comment]
In the past you did only ssl temporary because the heavy used of cpu and latency.
But now is common when you do ssl that you do a full ssl enviorment ->
is not more common to do only partial ssl stuff.

**Comment by colinschoen** [comment]
That is exactly my point... that is why the image proxy needs to not explode when nearly all forums are going to be using it because they are using SSL.

**Comment by albertlast** [comment]
But the proxy size can not grow unlimited?
Because at one point all images are cached,
so would be mean you env needs to be right sized to fitt your size.

**Comment by colinschoen** [comment]
^ Can you rephrase please.

**Comment by albertlast** [comment]
Well you forum got a specific numbers of images let guess, 1000 img.
After time mostly all pic are in your proxy,
beside the user add some images they got also than added to the proxy.
Which results in specific size of the proxy which is needed for the proxy to work.
After reaching this point,
the proxy size should grow naturaly only on the base when user add a image?
Because the proxy will not cache a image multiple time.

Which leads me to the idea that his proxy will not grow every day about 600mb,
it's only on the start and this behavior is essential to work for proxy -> no need change.

Btw the code say that the images get deleted after 5 days:
https://github.com/SimpleMachines/SMF2.1/blob/ac85d8c0fb1b070aaabfe622cf02839682c7eafb/proxy.php#L111-L117

**Comment by albertlast** [comment]
@sbulen what is the issue with existing solution to delete everything which is older than 5 days?

**Comment by sbulen** [comment]
@albertlast - I've never seen it work...

That logic appears to be in the middle of the serve logic, so I believe it will only kick in if you request the image again???  Which is a strange time to delete it...

(And it's followed by another serve, which is also confusing - it looks like it just deletes it & adds it again?  I'm not sure how that is intended to work...)   

In normal cache management, you should ideally go thru & get rid of the least recently used items somehow.  

I believe this won't be an issue for smaller sites, but on my 2.0 mid-sized site, the cache mushroomed to several GB within days, with no end in sight.  This halted our incremental backup process.  There are real operational issues with just adding several GB to a site...

I think we'll see more of these... Folks in 2.0 are deleting the cache by hand:
https://www.simplemachines.org/community/index.php?topic=558716.msg3960642#msg3960642

**Comment by albertlast** [comment]
Well this solution is a cache,
and a cached shoud also update the existing stuff,
so a definition of x days/timeunites is a good idea to refresh the existing stuff.

to know which is the least used you had to measure the use,
which could create again a overhead because you had to write this information somewhere.

The strong part of this implementation is that you got an easy why do finde the entries which are old,
by selecting the create date (filesystem operation) which are older than x days/timeunites.

When you safe the use of the entries you had to open every entry and look into the json information and read the information,
which create a big amount of overhead.

**Comment by sbulen** [comment]
Yes you are correct - this should emphatically *not* be an online function.

This should fall more into the category of regular maintenance, similar to our weekly attachment cleanup or other batch jobs.

You can't tell at retrieval time if something is too old - because you just retrieved it & the clock should start over again.

**Comment by Sesquipedalian** [comment]
> Btw the code say that the images get deleted after 5 days:

That bit of code is meant to trigger a periodic update of the proxy's stored copy, in case the image changed at the source. It doesn't do any maintenance.

**Comment by Sesquipedalian** [comment]
I agree that incorporating something like @sbulen's proxy-maint-cron.php into SMF's maintenance tasks is an appropriate solution. I'll look into what I think is the best way to implement it, unless @sbulen would rather tackle it himself.

**Comment by sbulen** [comment]
> > Btw the code say that the images get deleted after 5 days:
> 
> That bit of code is meant to trigger a periodic update of the proxy's stored copy, in case the image changed at the source. It doesn't do any maintenance.

D'oh!   (facepalm...)   NOW it makes sense...

I'm OK with taking it on, but it may take me a couple weeks.   Lots of RL challenges the last couple months.  The fog is just starting to clear...
